rbc/externals/__init__.py

```diff
@@ -29,37 +27,28 @@ def sanitize(name):
     return name
 
 
-def register_external(
-    fname,
-    retty,
-    argtys,
-    module_name,
-    module_globals,
-    typing_registry,
-    lowering_registry,
-    doc,
-):
-
-    # expose
-    fn = eval(f'lambda {",".join(map(lambda x: sanitize(x.name), argtys))}: None', {}, {})
-    _key = py_types.FunctionType(fn.__code__, {}, fname)
-    _key.__module__ = __name__
-    globals()[fname] = _key
-
-    # typing
-    @typing_registry.register_global(_key)
-    class ExternalTemplate(typing.templates.AbstractTemplate):
-        key = _key
-
-        def generic(self, args, kws):
-            retty_ = Type.fromobject(retty).tonumba()
-            argtys_ = tuple(map(lambda x: Type.fromobject(x.ty).tonumba(), argtys))
-            codegen = gen_codegen(fname)
-            lowering_registry.lower(_key, *argtys_)(codegen)
-            return retty_(*argtys_)
-
-    module_globals[fname] = _key
-    _key.__module__ = module_name
-    _key.__doc__ = doc
-    del globals()[fname]
-    return _key
+def make_intrinsic(fname, retty, argnames, module_name, module_globals, doc):
+    argnames = tuple(map(lambda x: sanitize(x), argnames))
+    fn_str = (
+        f'from numba.core.extending import intrinsic\n'
+        f'@intrinsic\n'
+        f'def {fname}(typingctx, {", ".join(argnames)}):\n'
+        f'    from rbc.typesystem import Type\n'
+        f'    retty_ = Type.fromobject("{retty}").tonumba()\n'
+        f'    argnames_ = tuple(map(lambda x: Type.fromobject(x).tonumba(), [{", ".join(argnames)}]))\n'  # noqa: E501
+        f'    signature = retty_(*argnames_)\n'
+        f'    def codegen(context, builder, sig, args):\n'
+        f'        from numba.core import funcdesc\n'
+        f'        fndesc = funcdesc.ExternalFunctionDescriptor("{fname}", sig.return_type, sig.args)\n'  # noqa: E501
+        f'        func = context.declare_external_function(builder.module, fndesc)\n'
+        f'        return builder.call(func, args)\n'
+        f'    return signature, codegen'
+    )
+
+    exec(fn_str)
+    fn = locals()[fname]
+    fn.__module__ = module_name
+    fn.__doc__ = doc
+    fn.__name__ = fname
+    module_globals[fname] = fn
```

Comment:
Nit: I wonder if this could be rewritten as:
```suggestion
    exec(fn_str, module_globals)
    fn = module_globals[fname]
    fn.__doc__ = doc
```
If yes, `module_name` can be eliminated, otherwise, never mind.